[CCLEX-173]: Health panel in Cluster page

AntonNos opened the pull request:
![ezgif-2-629ab89fea](https://user-images.githubusercontent.com/33761040/211051804-8ea99b08-1649-444f-9723-5ab1ddb39e2e.gif)

I made 3 fullwidth columns, because only on >1920px monitors it looks good with 50% columns.

**Update:** 

Light mode:
<img width="1792" alt="Screenshot 2023-01-12 at 21 25 23" src="https://user-images.githubusercontent.com/33761040/212206840-407529dc-5782-438f-9437-4686dc15e8bb.png">

No lma:
<img width="1788" alt="Screenshot 2023-01-13 at 02 05 01" src="https://user-images.githubusercontent.com/33761040/212206911-a08985d9-1db7-4630-9da1-8a2952f50d39.png">

**UPD: 50% width columns in Overview panel** 
<img width="1792" alt="Screenshot 2023-01-18 at 00 42 36" src="https://user-images.githubusercontent.com/33761040/213028580-78dd919d-f17a-4ef6-b2a2-e04c22cefbb9.png">

stefcameron commented:
👀

stefcameron commented on src/renderer/components/ClusterPage/Overview/MetricChart.js:
```diff
@@ -0,0 +1,73 @@
+
+ChartJS.register(ArcElement);
+
+const style = getComputedStyle(document.body);
```
I don't think you can do this just once when the module is loaded because I don't think that will work with theming. You need to do this every time something gets rendered, which is too bad because calling `getComputedStyle()` causes [layout reflow](https://gist.github.com/paulirish/5d52fb081b3570c81e3a#what-forces-layout--reflow), which means additional renders. Generally, it's best to avoid these APIs as much as possible. But I understand why you're using it: To avoid hard-coding color HEX values. It's probably fine since we're not calling it in a loop, which would be the worst case scenario.

But still, I think you need to call it _inside_ `getData()`, and `defaultData` should be a function `getDefaultData()` that calls it inside too, and returns the default data object...

stefcameron commented on src/renderer/components/ClusterPage/Overview/SingleMetric.js:
```diff
@@ -0,0 +1,82 @@
+            chartFillPercentage={chartFillPercentage}
+          />
+          <PercentageValue color={chartColor}>
+            {chartFillPercentage}%
```
I know it's just a percentage sign, but I'd prefer to keep all formatted strings coming from `strings.ts` as much as possible, so a strings.ts string that takes the `chartFillPercentage` and returns `"${chartFillPercentage}%"` would be my preference over hard coding the "%" string here.

AntonNos replied:
Makes sense, fixed

stefcameron commented on src/renderer/components/ClusterPage/Overview/HealthPanel.js:
```diff
@@ -1,7 +1,331 @@
+  return [
+    {
+      [`${strings.clusterPage.pages.overview.health.metrics.cpu.used()}`]: used
+        ? `${Math.round(used * 100)}%`
```
I know it's just a percentage sign, but I'd prefer to keep all formatted strings coming from `strings.ts` as much as possible, so a strings.ts string that takes the `chartFillPercentage` and returns `"${chartFillPercentage}%"` would be my preference over hard coding the "%" string here.

AntonNos replied:
Fixed

AntonNos commented on package.json:
```diff
@@ -61,7 +61,10 @@
   },
-  "dependencies": {},
+  "dependencies": {
+    "chart.js": "^4.1.1",
+    "react-chartjs-2": "^5.1.0"
```
Done

AntonNos commented on src/renderer/components/ClusterPage/Overview/ClusterOverviewView.js:
```diff
@@ -21,19 +21,25 @@ export const ClusterOverviewView = function ({ clusterEntity }) {
+        <div>
+          <SummaryPanel clusterEntity={clusterEntity} />
+        </div>
+        <div>
+          {FEAT_CLUSTER_PAGE_HEALTH_ENABLED ? (
+            <HealthPanel clusterEntity={clusterEntity} />
+          ) : null}
+        </div>
```
Right, thanks. Fixed

AntonNos commented on src/renderer/components/ClusterPage/Overview/ClusterOverviewView.js:
```diff
@@ -21,19 +21,25 @@ export const ClusterOverviewView = function ({ clusterEntity }) {
+          ) : null}
+        </div>
+        <div>
+          <DrawerTitleWrapper>
+            <PanelTitle
+              title={strings.clusterPage.pages.overview.clusterConditions.title()}
+            />
+          </DrawerTitleWrapper>
```
Hmm, I don't know why title is here😄 Fixed

AntonNos commented on src/renderer/components/ClusterPage/Overview/HealthPanel.js:
```diff
@@ -1,7 +1,331 @@
+
+/**
+ * Converts size in bytes to KB, MB, GB etc.
+ * @param {string|number} bytes The size in bytes.
```
Yes, this is much cleaner, thanks. Fixed

AntonNos commented on src/renderer/components/ClusterPage/Overview/HealthPanel.js:
```diff
@@ -1,7 +1,331 @@
+/**
+ * Creates array with objects of CPU metrics for future render.
+ * @param {Object} obj CPU metrics values.
+ * @param {string|number} obj.used Used value from CPU metrics.
+ * @param {string|number} obj.system System value from CPU metrics.
+ * @param {string|number} obj.io I/O value from CPU metrics.
+ * @param {string|number} obj.idle Idle value from CPU metrics.
```
Fixed, based on our discussion😉

AntonNos commented on src/renderer/components/ClusterPage/Overview/HealthPanel.js:
```diff
@@ -1,7 +1,331 @@
+ */
+const getCpuData = ({ used, system, io, idle }) => {
+  return [
+    {
+      [`${strings.clusterPage.pages.overview.health.metrics.cpu.used()}`]: used
+        ? `${Math.round(used * 100)}%`
+        : strings.catalog.entities.common.details.emptyValue(),
+    },
+    {
+      [`${strings.clusterPage.pages.overview.health.metrics.cpu.system()}`]:
+        system
+          ? `${Math.round(system * 100)}%`
+          : strings.catalog.entities.common.details.emptyValue(),
+    },
+    {
+      [`${strings.clusterPage.pages.overview.health.metrics.cpu.io()}`]: io
+        ? `${Math.round(io * 100)}%`
+        : strings.catalog.entities.common.details.emptyValue(),
+    },
+    {
+      [`${strings.clusterPage.pages.overview.health.metrics.cpu.idle()}`]: idle
+        ? `${Math.round(idle * 100)}%`
+        : strings.catalog.entities.common.details.emptyValue(),
+    },
```
Makes sense. Replaced the check from `used ? '' : ''` to `used >= 0 ? '' : ''`

AntonNos commented on src/renderer/components/ClusterPage/Overview/HealthPanel.js:
```diff
@@ -1,7 +1,331 @@
+ * @param {string|number} obj.system System value from CPU metrics.
+ * @param {string|number} obj.io I/O value from CPU metrics.
+ * @param {string|number} obj.idle Idle value from CPU metrics.
+ * @returns {Array<{ 'Used': string, 'System': string, 'I/O': string, 'Idle': string }>} array with objects of CPU metrics.
```
You are right, it will look much cleaner. Fixed.
Thanks for tip!

AntonNos commented on src/renderer/components/ClusterPage/Overview/HealthPanel.js:
```diff
@@ -1,7 +1,331 @@
+/**
+ * Creates array with objects of Memory metrics for future render.
+ * @param {Object} obj Memory metrics values.
+ * @param {string|number} obj.available Available memory value from Memory metrics.
+ * @param {string|number} obj.capacity Capacity value from Memory metrics.
+ * @param {string|number} obj.allocated Allocated memory value from Memory metrics.
+ * @returns {Array<{ 'Available': string, 'Capacity': string, 'Allocated': string }>} array with objects of Memory metrics.
```
Fixed

AntonNos commented on src/renderer/components/ClusterPage/Overview/HealthPanel.js:
```diff
@@ -1,7 +1,331 @@
+/**
+ * Creates array with objects of Storage metrics for future render.
+ * @param {Object} obj Storage metrics values.
+ * @param {string|number} obj.used Used storage value from Storage metrics.
+ * @param {string|number} obj.capacity Capacity value from Storage metrics.
+ * @param {string|number} obj.available Available storage value from Storage metrics.
+ * @returns {Array<{ 'Used': string, 'Capacity': string, 'Available': string }>} array with objects of Storage metrics.
```
Fixed

AntonNos commented on src/renderer/components/ClusterPage/Overview/HealthPanel.js:
```diff
@@ -1,7 +1,331 @@
+     * @param {Object} cloud Cloud.
+     */
+    const getMetrics = async (cloud) => {
+      const promUrl =
+        'https://a3f5c7cc1fc794891a8e86d6f29a8fe7-616423632.us-east-2.elb.amazonaws.com';
```
Ohh, I thought that this one is mocked. Fixed

AntonNos commented on src/renderer/components/ClusterPage/Overview/HealthPanel.js:
```diff
@@ -1,7 +1,331 @@
+  const [storagePercentage, setStoragePercentage] = useState(0);
+  const [timerTrigger, setTimerTrigger] = useState(0);
+
+  const isCpuFirstRender = useRef(true);
+  const isMemoryFirstRender = useRef(true);
+  const isStorageFirstRender = useRef(true);
```
You are right. We can use that case, when metrics are `null`. Thanks for tip! Fixed

AntonNos commented on src/renderer/components/ClusterPage/Overview/HealthPanel.js:
```diff
@@ -1,7 +1,331 @@
+      }
+
+      // Changes timer trigger every minute to update metrics.
+      setTimeout(() => {
+        setTimerTrigger(timerTrigger + 1);
+      }, UPDATE_METRICS_INTERVAL);
```
Done

AntonNos commented on src/renderer/components/ClusterPage/Overview/HealthPanel.js:
```diff
@@ -1,7 +1,331 @@
+ * @returns {string} `--` if no bytes provided, formatted size in other case.
+ */
+const formatBytes = (bytes, decimals = 2) => {
+  if (!bytes) {
```
Fixed

stefcameron commented:
👀

stefcameron commented on src/renderer/components/ClusterPage/clusterPageComponents.js:
```diff
@@ -76,3 +76,11 @@ export const TableSyncButton = styled.button(
 }));
+
+export const PanelsWrapper = styled.div(() => ({
+  overflow: 'auto',
```
Yesterday, I fixed an issue with scrolling by putting `overflow: 'auto',` on the PageContainer component. Should it be removed in light of this style here, or should this one be removed in favor of the one I added?

AntonNos replied:
Removed it from `PanelsWrapper` component. Having scrollbar in parent component is better I think. Thanks

stefcameron commented on src/renderer/components/ClusterPage/Overview/HealthPanel.js:
```diff
@@ -1,7 +1,360 @@
+    {
+      label: strings.clusterPage.pages.overview.health.metrics.cpu.used(),
+      value:
+        typeof used === 'number' && !Number.isNaN(used)
+          ? strings.clusterPage.pages.overview.health.metrics.cpu.percentageValue(
+              Math.round(used * 100)
+            )
+          : strings.catalog.entities.common.details.emptyValue(),
```
Since all 4 metrics are subject to the same checks and formatting, could this be moved into a small `formatPct()` function similar to `formatBytes()` that takes a `pct` value and does the checks and formatting? Then there's only one place to tweak if changes are needed, and the rest of the code just becomes like
```
{
   label: strings...,
   value: formatPct(system),
}
...
```

AntonNos replied:
Great idea, thanks for tip! Done.

stefcameron commented on src/renderer/components/ClusterPage/Overview/HealthPanel.js:
```diff
@@ -1,7 +1,360 @@
+     * @param {Object} cloud Cloud.
+     */
+    const getMetrics = async (cloud) => {
+      const promUrl = clusterEntity.spec.lma.prometheusUrl;
```
This is good, but we will have a case where `lma === null/undefined` because StackLight isn't installed, or also (maybe, just covering all possibilities) where `lma.prometheusUrl === null/undefined` because Prometheus isn't there.

In that case, we should be in this case: https://www.figma.com/file/6LYlfoeqjDMEdtKnhmoLgd/MCC%2FLens?node-id=56%3A2194&t=l8ZhpR6gvU80ZgbJ-4

<img width="1168" alt="image" src="https://user-images.githubusercontent.com/2855350/211084148-a60fa8ca-9f33-4de3-96f7-06ca79d52742.png">

AntonNos replied:
Oh, right. Sorry, I forgot about that case. Added it.
Thank you!

stefcameron commented on src/renderer/components/ClusterPage/Overview/HealthPanel.js:
```diff
@@ -1,7 +1,360 @@
+      const promUrl = clusterEntity.spec.lma.prometheusUrl;
+
+      if (promUrl) {
+        try {
+          const metrics = await getCpuMetrics(cloud, promUrl);
+          setCpuMetrics(metrics);
+        } catch (err) {
+          logger.error(err.message);
+          setCpuMetrics({});
+        }
+
+        try {
+          const metrics = await getMemoryMetrics(cloud, promUrl);
+          setMemoryMetrics(metrics);
+        } catch (err) {
+          logger.error(err.message);
+          setMemoryMetrics({});
+        }
+
+        try {
+          const metrics = await getDiskMetrics(cloud, promUrl);
+          setStorageMetrics(metrics);
+        } catch (err) {
+          logger.error(err.message);
+          setStorageMetrics({});
+        }
```
Instead of executing these requests serially, it would be faster to do them in parallel with `Promise.allSettled()` (which will give you results for each whether they succeed or fail so you can still show individual error messages).

And you should really test the error cases. `logger.error()` expects a "scope" as its first parameter, and then a message as its second, and will throw if it only gets one parameter... Busted! 😉 

Since this is running inside an effect, you have to get a bit creative. This is what I would suggest. The point is using a scope that would help you quickly identify _where_ this log message came from so you can quickly find the spot where the error occurred.

```javascript
import { logger, logValue } from '../util/logger';
...
logger.error(
  'HealthPanel.useEffect.getMetrics()',
  `Failed to get CPU metrics; error=${logValue(err)}`
);
```

AntonNos replied:
Great idea. It really works a lot faster that serially requests 🔥
Done

stefcameron commented on src/renderer/components/ClusterPage/Overview/HealthPanel.js:
```diff
@@ -1,7 +1,360 @@
+  }, [clouds, clusterEntity, timerTrigger]);
+
+  useEffect(() => {
+    if (cpuMetrics !== null) {
+      if (cpuMetrics) {
```
Since the initial state of `cpuMetrics` is `null`, I'm struggling to find a case where the `else` statement on line 265 would ever get called. First pass, `cpuMetrics === null` so this effect does nothing. Meanwhile, the effect above is executing and waiting for the async request to get CPU metrics to complete. Request completes, whether successful or not, we have `cpuMetrics !== null`. Then this effect runs again because `cpuMetrics` has changed, and then both lines 251 and 252 succeed and we fall into the `if` block.

Also, if that `else` block were ever to be called, it would throw because `getCpuData()`'s signature is `getCpuData({ ... })` which means it expects the first parameter to be an object it can destructure, yet you're passing `undefined`. That will cause an exception. You'd had to define `getCpuData` as `getCpuData({ ... } = {})` to allow it to be called as `getCpuData()` or `getCpuData({})` or `getCpuData({ ... })`.

But I don't think that's necessary, because the effect above that makes the async calls handles failures and will always set `cpuMetrics` to _something_ that isn't `null`.

So I think you can eliminate line 251 as well as the `else` block at line 264.

And the same goes for memory and storage effects below.

AntonNos replied:
Hmmm, you are right, we don't need that `if (cpuMetrics !== null)` and else block here. Thanks for tip

stefcameron commented on src/renderer/components/ClusterPage/Overview/HealthPanel.js:
```diff
@@ -1,7 +1,360 @@
+
+  // Canceling the timeout if the component gets unmounted
+  useEffect(() => {
+    mounted.current = true;
```
What is this ref used for?

AntonNos replied:
It checks if component was mounted. Found that here -> https://stackoverflow.com/questions/39767482/is-there-a-way-to-check-if-the-react-component-is-unmounted

stefcameron replied:
OK, but what's the benefit of knowing this? So far, I'm not seeing how that helps anything. 🙂

AntonNos replied:
You are right. We don't need it:)
Thanks, removed

stefcameron commented on src/renderer/components/ClusterPage/Overview/HealthPanel.js:
```diff
@@ -1,7 +1,360 @@
+  useEffect(() => {
+    mounted.current = true;
+
+    return () => {
+      clearTimeout(updateMetricsTimeoutId);
+      mounted.current = false;
+    };
```
I was expecting this to be in the effect where you're making the async calls and setting the timeout. You _probably_ don't need to keep the timer ID in state either because you're already using another state variable (`timerTrigger`) to cause the timeout to be re-established when it expires.

AntonNos replied:
Fixed

stefcameron commented on src/strings.ts:
```diff
@@ -382,6 +382,64 @@ export const clusterPage: Dict = {
+            available: () => 'Available',
+            tooltipInfoHtml: () => `
+              <p>Metrics are received from the kubernetes cluster and may differ from OS level metrics.</p>
+              <p>Capacity: Total disk space across.</p>
+              <p>Available: Total free disk space.</p>
+              <p>Used: Total consumed disk space.</p>
```
These strings are all missing "across the cluster".

AntonNos replied:
I asked @john-papadami about text strings on Tuesday, and he confirmed, that actual text strings are in Figma.
I leave them as it is for now. If it is necessary to change it, will do that. Just let me know what to do with text.
Thank you

stefcameron commented on src/strings.ts:
```diff
@@ -382,6 +382,64 @@ export const clusterPage: Dict = {
+            system: () => 'System',
+            io: () => 'I/O',
+            idle: () => 'Idle',
+            tooltipInfoHtml: () => `
+              <p>Metrics are retrieved from the kubernetes cluster and may differ from OS level metrics.</p>
+              <p>Used: 1-minute average time spent performing.</p>
+              <p>System: 1-minute average time spent in the kernel.</p>
+              <p>I/O: 1-minute average time spent waiting for I/O.</p>
+              <p>Idle: 1-minute average idle time not performing tasks.</p>
+            `,
+            percentageValue: (percentage) => `${percentage}%`,
+          },
+          memory: {
+            title: () => 'Memory',
+            available: () => 'Available',
+            capacity: () => 'Capacity',
+            allocated: () => 'Allocated',
+            tooltipInfoHtml: () => `
+              <p>Metrics are retrieved from the kubernetes cluster and may differ from OS level metrics.</p>
+              <p>Available: Total availability for starting new applications, without the need for swapping.</p>
+              <p>Capacity: Total memory including available and allocated.</p>
+              <p>Allocated: Total memory currently in use by both the OS and running applications.</p>
+            `,
+          },
+          storage: {
+            title: () => 'Storage',
+            used: () => 'Used',
+            capacity: () => 'Capacity',
+            available: () => 'Available',
+            tooltipInfoHtml: () => `
+              <p>Metrics are received from the kubernetes cluster and may differ from OS level metrics.</p>
+              <p>Capacity: Total disk space across.</p>
+              <p>Available: Total free disk space.</p>
+              <p>Used: Total consumed disk space.</p>
```
Actually, most of these strings aren't what is in https://mirantis.jira.com/browse/CCLEX-173 at the "Metric info explanations" section. 🤔

stefcameron replied:
I see. John and I didn't expressly "sync" our strings between the ticket and Figma.

My main concern with the strings describing the metrics is that they don't feel precise enough to differentiate them from what users already see in Lens in terms of CPU, Memory, and Storage. I think users will look at this and wonder why we put this in -- and worse, that we're probably getting it wrong because our metrics won't look like the ones in Lens.

But also, many of the strings, as they are now in the code, are missing words. They are incomplete. Either those are typos, or...? Like "Used: 1-minute average time spent performing." -- performing what? Or 
Capacity: Total disk space across." -- across what?

I think the strings in the ticket are more complete.

AntonNos replied:
Got it, will change them to be similar with ticket.

stefcameron reviewed (changes requested):
Thank you for those improvements!

stefcameron commented:
> I made 3 fullwidth columns, because only on >1920px monitors it looks good with 50% columns.

Really? It's not possible to make the graphs smaller (and so narrower) to give in a side-by-side panel with Overview?

Also, you know I'm going to ask... What about light mode? 😉

stefcameron commented on src/strings.ts:
```diff
@@ -382,6 +382,72 @@ export const clusterPage: Dict = {
+            io: () => 'I/O',
+            idle: () => 'Idle',
+            tooltipInfoHtml: () => `
+              <p>Metrics are retrieved from the kubernetes cluster and may differ from OS level metrics.</p>
+              <p>Used (%): 1-minute average time spent performing tasks, across the cluster.</p>
+              <p>System (%): 1-minute average time spent in the kernel, across the cluster.</p>
+              <p>I/O (%): 1-minute average time spent waiting for I/O, meaning system was disk- or network-bound, across the cluster.</p>
+              <p>Idle (%): 1-minute average idle time not performing tasks, across the cluster.</p>
```
I had a look at the tooltip in Lens. I think we could improve on it a bit to make it a little more clear by making a list out of the items, and then putting the disclaimed at the bottom with an asterisk (*) next to it, like this:

---

\- Used (%): 1-minute average time spent performing tasks, across the cluster.
\- System (%): 1-minute average time spent in the kernel, across the cluster.
\- I/O (%): 1-minute average time spent waiting for I/O, meaning system was disk- or network-bound, across the cluster.
\- Idle (%): 1-minute average idle time not performing tasks, across the cluster.

\* Metrics are retrieved from the kubernetes cluster and may differ from OS level metrics.

---

```
<p>- Used (%): 1-minute average time spent performing tasks, across the cluster.</p>
<p>- System (%): 1-minute average time spent in the kernel, across the cluster.</p>
<p>- I/O (%): 1-minute average time spent waiting for I/O, meaning system was disk- or network-bound, across the cluster.</p>
<p>- Idle (%): 1-minute average idle time not performing tasks, across the cluster.</p>
<p class="disclaimer">* Metrics are retrieved from the kubernetes cluster and may differ from OS level metrics.</p>
```

And then style the `p.disclaimer` class in MetricTitle.js's `<Title>` styled component to have an 8px top margin so it looks like this:

<img width="730" alt="image" src="https://user-images.githubusercontent.com/2855350/212392868-ad07c21f-0fdc-4a14-bb43-9114dd098fd0.png">

Same for the memory and storage tooltips.

AntonNos replied:
Nice idea, thank you for tip about tooltip😄
Fixed. Also fixed width of columns in Overview panel.

stefcameron commented:
> > I made 3 fullwidth columns, because only on >1920px monitors it looks good with 50% columns.
> 
> Really? It's not possible to make the graphs smaller (and so narrower) to give in a side-by-side panel with Overview?

Right now, I can shrink the Lens window (making it narrower) and the health panel responds, including the graphs:

<img width="566" alt="image" src="https://user-images.githubusercontent.com/2855350/212393304-0e8db80a-13ce-403f-a39d-308950bebe1d.png">

The graphs still look good. Why can't we have the Health panel 50% on the right, with Overview 50% on the left, like in the mockups?

stefcameron reviewed (commented):
Looking good! I just have a small request to tweak the tooltips, and then still wondering why the health panel can't be narrower -- see my previous comment in the PR.

stefcameron reviewed (commented):
Visually, all the changes look great now!

But, I discovered that if I switch from dark to light mode, or vice versa, the background color of the charts doesn't switch.

For example, when going from dark to light:

<img width="609" alt="Screenshot 2023-01-17 at 5 29 14 PM" src="https://user-images.githubusercontent.com/2855350/213035164-8712d649-4748-4c5c-a156-1b65001dc4e9.png">
<img width="610" alt="Screenshot 2023-01-17 at 5 29 35 PM" src="https://user-images.githubusercontent.com/2855350/213035165-e8496b69-3ff2-41f6-be82-9645ef81211a.png">

The same problem occurs if you start out in light mode and then change to dark mode. The background color will remain the one in light mode.

AntonNos commented:
> Visually, all the changes look great now!
> 
> But, I discovered that if I switch from dark to light mode, or vice versa, the background color of the charts doesn't switch.
> 
> For example, when going from dark to light:
> 
> <img alt="Screenshot 2023-01-17 at 5 29 14 PM" width="609" src="https://user-images.githubusercontent.com/2855350/213035164-8712d649-4748-4c5c-a156-1b65001dc4e9.png">
> 
> <img alt="Screenshot 2023-01-17 at 5 29 35 PM" width="610" src="https://user-images.githubusercontent.com/2855350/213035165-e8496b69-3ff2-41f6-be82-9645ef81211a.png">
> 
> The same problem occurs if you start out in light mode and then change to dark mode. The background color will remain the one in light mode.

Hmm, didn't see that. 
Fixed it by adding `theme.mode` to `useEffect` dependencies.

stefcameron commented on src/renderer/components/ClusterPage/Overview/MetricChart.js:
```diff
@@ -63,10 +64,11 @@ const options = {
   useEffect(() => {
     setChartData(getData(chartColor, chartFillPercentage));
-  }, [chartColor, chartFillPercentage]);
+  }, [chartColor, chartFillPercentage, theme.mode]);
```
In theory, this seems good, but when I run it, while the colors to update, the chart data doesn't load (or remains at zero, or gets reset to zero, not sure):

<img width="592" alt="image" src="https://user-images.githubusercontent.com/2855350/213212632-032dd4e2-77fc-4311-a585-c35fe02b6e8a.png">

AntonNos replied:
Fixed. Added new error message and reconnect button.

stefcameron commented on src/strings.ts:
```diff
@@ -385,13 +385,20 @@ export const clusterPage: Dict = {
+            disconnectedManagementCluster: {
+              title: () =>
+                'Health metrics cannot be displayed for this cluster because its management cluster is disconnected.',
+              reconnectButtonLabel: () => 'Reconnect!',
```
I think just "Reconnect" without the exclamation mark (!) is sufficient.

AntonNos replied:
Fixed

stefcameron reviewed (changes requested):
Thank you for the update! In general, this looks good, but it's not really working well.

When disconnected, I see this:

<img width="589" alt="Screenshot 2023-01-20 at 9 29 23 AM" src="https://user-images.githubusercontent.com/2855350/213737643-58e0b5f0-921a-463c-84a5-779cdde10288.png">

The paragraph should be up next to the error icon, and the Reconnect link should be right at the end of the text, not on the next line.

Also, when I click Reconnect, the spinners are a nice touch, but they never go away. I stays like this forever:

<img width="590" alt="Screenshot 2023-01-20 at 9 30 43 AM" src="https://user-images.githubusercontent.com/2855350/213738253-7a9d9ade-bc20-45ed-8d6e-63caf0ea570c.png">

AntonNos commented:
> Thank you for the update! In general, this looks good, but it's not really working well.
> 
> When disconnected, I see this:
> 
> <img alt="Screenshot 2023-01-20 at 9 29 23 AM" width="589" src="https://user-images.githubusercontent.com/2855350/213737643-58e0b5f0-921a-463c-84a5-779cdde10288.png">
> 
> The paragraph should be up next to the error icon, and the Reconnect link should be right at the end of the text, not on the next line.
> 
> Also, when I click Reconnect, the spinners are a nice touch, but they never go away. I stays like this forever:
> 
> <img alt="Screenshot 2023-01-20 at 9 30 43 AM" width="590" src="https://user-images.githubusercontent.com/2855350/213738253-7a9d9ade-bc20-45ed-8d6e-63caf0ea570c.png">

Fixed. Now we have a spinner while `isCloudFetching = true` and once it updates, charts and info updates too

stefcameron reviewed (changes requested):
Thanks! It's better, but there are still some issues:

<img width="664" alt="Screenshot 2023-01-23 at 7 19 05 PM" src="https://user-images.githubusercontent.com/2855350/214193723-4fb9bd5b-ed8a-4a53-8751-2838bd0f2081.png">

1. The text shouldn't go under the error icon. Usually this means a flex container with an icon and a paragraph, and the icon uses `flex: none` and paragraph `flex: 1` (so it can grow/shrink). That should result in the paragraph not flowing under the icon.
2. The CPU, Memory, and Storage labels don't appear centered within their columns with respect to their graphs.
3. I somehow got into this state where the bottom of the health panel was missing its gap. Not sure how. Can't always repro.

Finally, there's what we call "jank" when reconnecting (see how it flashes back to "disconnected" error and back to spinners while it's reconnecting):

![cclex-jank](https://user-images.githubusercontent.com/2855350/214194089-74949eb1-cd5f-499e-9077-03dad12ec4e0.gif)

I think you need a local state variable to put yourself into spinner mode as soon as the link is clicked rather than relying solely on the cloud events.

Just make sure to come back out of it if the reconnection fails... We should end up back at this disconnected error if reconnection fails.

stefcameron commented:
@AntonNos About

> The text shouldn't go under the error icon. Usually this means a flex container with an icon and a paragraph, and the icon uses flex: none and paragraph flex: 1 (so it can grow/shrink). That should result in the paragraph not flowing under the icon.

Maybe you can use our little `<InlineNotice>` component...?

AntonNos commented:
> @AntonNos About
> 
> > The text shouldn't go under the error icon. Usually this means a flex container with an icon and a paragraph, and the icon uses flex: none and paragraph flex: 1 (so it can grow/shrink). That should result in the paragraph not flowing under the icon.
> 
> Maybe you can use our little `<InlineNotice>` component...?

Thanks for tip! I will update PR today and will try to use `<InlineNotice>` here.

AntonNos commented:
> Thanks! It's better, but there are still some issues:
> 
> <img alt="Screenshot 2023-01-23 at 7 19 05 PM" width="664" src="https://user-images.githubusercontent.com/2855350/214193723-4fb9bd5b-ed8a-4a53-8751-2838bd0f2081.png">
> 
> 1. The text shouldn't go under the error icon. Usually this means a flex container with an icon and a paragraph, and the icon uses `flex: none` and paragraph `flex: 1` (so it can grow/shrink). That should result in the paragraph not flowing under the icon.
> 2. The CPU, Memory, and Storage labels don't appear centered within their columns with respect to their graphs.
> 3. I somehow got into this state where the bottom of the health panel was missing its gap. Not sure how. Can't always repro.
> 
> Finally, there's what we call "jank" when reconnecting (see how it flashes back to "disconnected" error and back to spinners while it's reconnecting):
> 
> ![cclex-jank](https://user-images.githubusercontent.com/2855350/214194089-74949eb1-cd5f-499e-9077-03dad12ec4e0.gif) [ ![cclex-jank](https://user-images.githubusercontent.com/2855350/214194089-74949eb1-cd5f-499e-9077-03dad12ec4e0.gif) ](https://user-images.githubusercontent.com/2855350/214194089-74949eb1-cd5f-499e-9077-03dad12ec4e0.gif) [ ](https://user-images.githubusercontent.com/2855350/214194089-74949eb1-cd5f-499e-9077-03dad12ec4e0.gif)
> 
> I think you need a local state variable to put yourself into spinner mode as soon as the link is clicked rather than relying solely on the cloud events.
> 
> Just make sure to come back out of it if the reconnection fails... We should end up back at this disconnected error if reconnection fails.

Fixed! Now it works much faster than before. Also used `InlineNotice` component for warning message, fixed bug with gap and centered title with tooltip icon.

stefcameron reviewed (changes requested):
Thank you for the update!

We're really, really close now. But my eye, it just keeps see things. 😉 

It bugs me that the error icon isn't centered with the first line of text:

<img width="551" alt="Screenshot 2023-01-25 at 7 08 18 PM" src="https://user-images.githubusercontent.com/2855350/214733015-7367a25c-989f-411b-80b8-7e561ebe8314.png">

All that's needed here is to fix `InlineNotice.js` like this at line 34:

```
const Container = styled.div(function () {
  return {
    display: 'flex',

    '> .Icon': {
      marginTop: 2,
    }
  };
});
```

(i.e. move the icon down 2px)

Then we get this (which also works with the other icons related to the `type` prop, like WARNING, INFO, NOTE, etc.):

<img width="558" alt="Screenshot 2023-01-25 at 7 10 02 PM" src="https://user-images.githubusercontent.com/2855350/214733199-0f7fb04e-ff82-4679-882b-35328ef29b1c.png">

And last but not least, if the Cloud is disconnected, but the cluster was already synced, and the cluster's `lma` property is defined, but for some reason, `lma.prometheus` is not (i.e. StackLight is there, but not Prom, for some reason, not sure if it's possible, but when dealing with MCC-related things, I not to make too many assumptions...), then I get this:

<img width="574" alt="Screenshot 2023-01-25 at 7 11 53 PM" src="https://user-images.githubusercontent.com/2855350/214733483-d5a5b524-003e-478e-b94b-42910bd98ae7.png">

Both errors show-up at the same time. The StackLight error should take precedence over the disconnected error because even if we were reconnected, we still wouldn't have data.

stefcameron commented:
> The StackLight error should take precedence over the disconnected error because even if we were reconnected, we still wouldn't have data.

With your fix, when we have both conditions as the same time (disconnected and no Prom URL), we see the "disconnected" error message -- but what we should see is the "no StackLight" error message, because even if they reconnected, we still wouldn't be able to show anything...

If they're disconnected _with_ a Prom URL, then we should see the "disconnected" message.

Makes sense?

AntonNos commented:
> > The StackLight error should take precedence over the disconnected error because even if we were reconnected, we still wouldn't have data.
> 
> With your fix, when we have both conditions as the same time (disconnected and no Prom URL), we see the "disconnected" error message -- but what we should see is the "no StackLight" error message, because even if they reconnected, we still wouldn't be able to show anything...
> 
> If they're disconnected _with_ a Prom URL, then we should see the "disconnected" message.
> 
> Makes sense?

I'm showing a message with list of problems without "Reconnect" button. Is that wrong message for that case (in both conditions in the same time)?

stefcameron commented on src/renderer/components/ClusterPage/Overview/HealthPanel.js:
```diff
@@ -350,6 +350,21 @@ export const HealthPanel = ({ clusterEntity }) => {
       <DrawerTitleWrapper>
         <PanelTitle title={strings.clusterPage.pages.overview.health.title()} />
       </DrawerTitleWrapper>
+      {(isNoMetrics && cloudStatus === CONNECTION_STATUSES.DISCONNECTED) ||
+        (isNoMetrics && (
```
The logic is a bit hard to understand here because there seem to be overlapping conditions related to `isNoMetrics === true`.

Feels like it should be `{isNoMetrics ? <ErrorMessage ... /> : null}` (BTW, it's a better practice to use the ternary operator and return `null` instead of doing `isNoMetrics && <ErrorMessage ...` because the former really results in nothing in the DOM while the latter `&&` case still affects the DOM)

And then line 368 should be `{!isNoMetrics && cloudStatus === CONNECTION_STATUSES.DISCONNECTED ? ... : null}`

BTW, double-negatives are really hard to process in the human brain and you should always try to avoid them as much as possible.

It's always better to find a positive way to name variables so you don't end up with cases like `!isNoMetrics` -- what? Oh, "not-no-metrics", so we _have_ metrics. That's a lot harder to process than `!haveMetrics` or `!promUrl` -- "I don't have metrics" or "No Prom URL". 🙂 

Like I wish HTML had never come up with a `disabled` attribute! That's a double-negative, because you inevitably come up to cases where you have to test, `if (!disabled)` -- much harder to reason in your brain than, `if (enabled)`, right?

AntonNos replied:
You are right, bad naming is hard to understand. I'm trying to name variables logically, in that case I couldn't think of a better name that "isNoMetrics"🤷‍♂️
Will take into account about ternary operator instead of `... && ()`.
Thank you for tips!
I'll make the changes in a few hours, because I'm not near computer right now.

stefcameron commented:
> 

This is what I see when I'm both disconnected _and_ don't have prometheus on a cluster:

<img width="600" alt="image" src="https://user-images.githubusercontent.com/2855350/215150377-a3955831-b86a-4e0c-9659-fa013d497416.png">

I should be seeing the other message instead.

AntonNos commented:
> > > > The StackLight error should take precedence over the disconnected error because even if we were reconnected, we still wouldn't have data.
> > > 
> > > 
> > > With your fix, when we have both conditions as the same time (disconnected and no Prom URL), we see the "disconnected" error message -- but what we should see is the "no StackLight" error message, because even if they reconnected, we still wouldn't be able to show anything...
> > > If they're disconnected _with_ a Prom URL, then we should see the "disconnected" message.
> > > Makes sense?
> > 
> > 
> > I'm showing a message with list of problems without "Reconnect" button. Is that wrong message for that case (in both conditions in the same time)?
> 
> This is what I see when I'm both disconnected _and_ don't have prometheus on a cluster:
> 
> <img alt="image" width="600" src="https://user-images.githubusercontent.com/2855350/215150377-a3955831-b86a-4e0c-9659-fa013d497416.png">
> 
> I should be seeing the other message instead.

Fixed, now we have correct message. I think the problem was because of `... && ()`. Now it works good. Thanks again for tip!

stefcameron commented on src/renderer/components/ClusterPage/Overview/HealthPanel.js:
```diff
@@ -350,22 +350,22 @@ export const HealthPanel = ({ clusterEntity }) => {
-          </ErrorMessage>
-        ))}
-      {cloudStatus === CONNECTION_STATUSES.DISCONNECTED && (
+      {(!hasPromUrl && cloudStatus === CONNECTION_STATUSES.DISCONNECTED) ||
+      !hasPromUrl ? (
```
Because of the `&&` operator here, and the fact that on both sides of the `||` you're testing the same condition `!hasPromUrl`, the entire expression can be reduced to just `!hasPromUrl`.

If we don't have a prom URL, then the connection status really doesn't matter: We need to show we can't get the data because the StackLight config doesn't support it.

And then line 368 is good because it read, "if we do have a prom url, and we're disconnected, then show the notice about having no data because we're disconnected and offer to reconnect."

I tested the above locally to be 100% sure.

AntonNos replied:
You are right, we don't care about connection status if we don't have a prom URL. Fixed

stefcameron reviewed (changes requested):
That's it. The last thing.

stefcameron reviewed (approved):
🚀 Thank you for sticking with it to the end and getting all those smaller issues fixed too!

AntonNos commented:
> 🚀 Thank you for sticking with it to the end and getting all those smaller issues fixed too!

Thank you for all the tips and comments!
